### PR TITLE
RDRP-1050 RDRP-1072 MoR module works for PNP

### PR DESCRIPTION
**Pull request by AnneONS**
## Pull Request submission

imputation_main.py has conditions so only MoR runs if the pipeline is in PNP mode. (RDRP-1050)

the MoR script is updated so short forms are not imputed in PNP mode, this is done using a new function returning a bool for long forms only. This function has a unit test. (RDRP-1072)

The logic in pipeline.py has changed as we now do include imputation in PNP mode, and this has also resulted in changes in apportionment_main.py

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #1050 and #1072

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `requirements.txt`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

**Review by Ryan2Y79 — CHANGES_REQUESTED** [review]
As a first check with this PR I ran `python main.py` and the pipeline dies (screenshot below). I have PNP and 2023 set in my `user_config`. Do you get the same error. If not, please call me. Cheers.

image

File: helpers/create_pnp_backdata.py
Comment (Ryan2Y79) [inline helpers/create_pnp_backdata.py:1]: Can we ensure that we want merged into develop the version of this script that exists on `RDRP-1081_pnp_2021_backdata` and not this branch.

Comment (AnneONS) [reply helpers/create_pnp_backdata.py:1]: I can think of 3 options: 

1.  revert this script to the develop version (a bit cumbersome) 
2. keep it as it is (it works) and improve it in the next merge of the other ticket
3. update it now to the version in the other PR 
what do you suggest?

File: src/imputation/MoR.py [hunk 1]
```diff
@@ -123,7 +86,7 @@ def mor_preprocessing(df, backdata, is_2022, config):
     return to_impute_df, remainder_df, backdata
 
 
-def carry_forwards(df, backdata, impute_vars):
+def carry_forwards(df, backdata, impute_vars, config):
     """Carry forwards matcing `backdata` values for references to be imputed.
```
Comment (Ryan2Y79) [inline src/imputation/MoR.py]: typo: matcing

**Review by Ryan2Y79 — APPROVED** [review]
Looks all good to me. Just one typo to fix. Thank you for adding in the unit tests to ensure function behaviours is as one expects.

File: helpers/create_pnp_backdata.py
Comment (Ryan2Y79) [reply helpers/create_pnp_backdata.py:1]: lets do option 3, so they are uniform with each other.

File: src/imputation/MoR.py (hunk 1 above)
Comment (AnneONS) [reply src/imputation/MoR.py]: done

**Pull request merged**